cpp/src/arrow/ipc/writer.cc

```diff
@@ -154,6 +154,10 @@ class RecordBatchSerializer {
       return Status::CapacityError("Cannot write arrays larger than 2^31 - 1 in length");
     }
 
+    if (arr.offset() != 0 && arr.device_type() != DeviceAllocationType::kCPU) {
+      return Status::NotImplemented("Cannot compute null count for non-cpu sliced array");
```

Comment:
Can we link this to a follow up ticket?

Reply:
I created a ticket https://github.com/apache/arrow/issues/43029 and added a comment here to link to it